gdal/swig/python/gdal-utils/osgeo_utils/auxiliary/color_palette.py

```diff
@@ -118,43 +121,89 @@ def assign(self, other: 'ColorPalette'):
+            if override and ('nv' in self.pal):
+                del self.pal['nv']
+
+    def read(self, filename_or_strings: Optional[ColorPaletteOrPathOrStrings]):
+        if filename_or_strings is None:
             self.pal.clear()
-            return self
-        elif base.is_path_like(color_filename_or_lines):
-            color_filename_or_lines = open(color_filename_or_lines).readlines()
-        elif not isinstance(color_filename_or_lines, Sequence):
-            raise Exception('unknown input {}'.format(color_filename_or_lines))
+            return
+        elif isinstance(filename_or_strings, ColorPalette):
+            self.assign(filename_or_strings)
+        elif base.is_path_like(filename_or_strings):
+            self.read_file(filename_or_strings)
+        elif isinstance(filename_or_strings, Sequence):
+            self.read_file_txt(lines=filename_or_strings)
+        else:
+            raise Exception('Unknown input {}'.format(filename_or_strings))
+
+    def read_file(self, filename: PathLikeOrStr):
+        ext = base.get_extension(filename).lower()
+        if ext in ['qlr', 'qml']:
+            self.read_file_qml(filename)
+        else:
+            self.read_file_txt(filename)
+
+    def read_file_qml(self, qml_filename, tag_name=None, type=None):
+        """ Read QGIS Layer Style File (qml) or QGIS Layer Definition File (qlr) """
+        qlr = minidom.parse(str(qml_filename))
+        if tag_name is None:
+            if type is None:
+                renderer = qlr.getElementsByTagName('rasterrenderer')
+                if renderer is None:
+                    raise Exception(f'Cannot find "rasterrenderer" in {qml_filename}')
+                type = renderer[0].getAttribute("type")
+                type_to_tag_name = {
+                    # <rasterrenderer type="paletted" opacity="1" alphaBand="-1" band="1" nodataColor="">
+                    # <paletteEntry color="#ffffff" alpha="0" label="0" value="0"/>
+                    "paletted": "paletteEntry",
+                    # <rasterrenderer type="singlebandpseudocolor" opacity="1" alphaBand="-1" band="1" classificationMax="100" classificationMin="0" nodataColor="">
+                    # <item label="-373" color="#d7191c" alpha="255" value="-373"/>
+                    "singlebandpseudocolor": "item",
+                }
+                if type not in type_to_tag_name:
+                    raise Exception(f'Unknown type: {type} in {qml_filename}')
+                tag_name = type_to_tag_name[type]
 
         self.pal.clear()
-        for line in color_filename_or_lines:
+        color_palette = qlr.getElementsByTagName(tag_name)
+        for palette_entry in color_palette:
+            color = palette_entry.getAttribute("color")
+            if str(color).startswith('#'):
+                color = int(color[1:], 16)
+            alpha = palette_entry.getAttribute("alpha")
+            alpha = int(alpha)
+            color = color + (alpha << 8 * 3)  # * 256**3
```

Comment:
This would look more idiomatic, at least to me, with 
```suggestion
            color = color + (alpha << 24)
```
(or at the very least add parenthesis around 8 * 3 to make it a no-brainer regarding operator precedence)